### ISSUE TITLE
modification to mu API

### ISSUE THREAD

**Issue by pelger** [apparatus/mu#71]
Hey @davidmarkclements, @mcollina,
Just reviewing the mu code so wanted to flag this one. It seems to me that the API is a little inconsistent. namely mu.define receives the entire message (pattern and protocol) whilst the final callback handler receives err and result. IMO the following might be better:

mu.define({...}, function(args, cb, msg) 

where args = msg.pattern
cb as before 
msg = full message (i.e. pattern + protocol)

mu.dispatch({...}, function(err, result, msg) 

where result = msg.response
msg = full message (i.e. response + protocol)

this way the msg can be ignored in most cases but is available for inspection if required.

Thoughts?

**Comment by mcollina** [comment]
👍 for dispatch.

As for `define`, maybe an option? or maybe attach the full message to `cb`?

**Comment by pelger** [comment]
I know it kinda breaks the canonical API pattern a little, but in most cases one would use:

mu.define({...}, args, cb)

as the msg structure is not normally required, it's just there 'in case' - Yep we could attach it to cb also but this feels a little hacky? Happy to go with the majority verdict however :) - @davidmarkclements - thoughts?

**Comment by davidmarkclements** [comment]
Yes - this is related to https://github.com/apparatus/mu/issues/43

Here's my thoughts

```
mu.dispatch(pattern, cb, optionalMsgStructureForSpecialCases)
cb: (err, args: {pattern: [non-enumerable, circular to parent (args)], protocol: [non-enumerable]})

mu.define(pattern, cb, optionalMsgStructureForSpecialCases)
cb: (err, result: {args: [non-enumerable, circular to parent (result)], protocol: [non-enumerable]})
```

benefits: 

* non-breaking change
* clarity for regular users
* avoid performance issues later on (no polymorphism)

drawbacks: 
* makes core a little more complex from a grokking point of view, but adds very little complexity to actual code - no extra layers etc. we just need to be careful with serialization when we *do* want to include the protocol object

**Comment by mcollina** [comment]
IMHO we still have a small userbase, and it's a good time for breaking changes.

This implementation will run easily in the browsers, too. I think touching `__proto__`  is not a good idea, because the `protocol` could then be overwritten by the end users. 

Exposing metadata is super easy to implement this using `Symbols`, but it won't be available in browsers.

```js
var obj = { hello: 'world' }
obj[Symbol.for('protocol')] = 'whaat?'
console.log(obj)
console.log(obj[Symbol.for('protocol')])
```

If we can feature detect their presence, and _turn off this feature_ for older browsers, we are god. `Symbol` is not available on Android < 5.1 and IE 11. I think we should go for this solution.

**Comment by pelger** [comment]
Following discussions gone for the simple solution of msg as last parameter to define and dispatch cbs.
